### DISCUSSION TITLE
Add feature for TeXLive Manager and outdated packages

### DISCUSSION THREAD

**Discussion by ghost** [FelixKratz/SketchyBar#525] [Q&A]
Very similar to the `homebrew`-module I like to have an overview of outdated TeXLive packages.

I added two files in `items` and `plugins` folder and made them executable. But the number is always `0` and therefore it says, no outdated package:

<img width="551" alt="Screenshot 2024-03-23 at 16 15 00" src="https://github.com/FelixKratz/SketchyBar/assets/12896078/ce73a9a7-2380-49e9-8e04-ab4fc26061da">

The `tlmgr.sh` in the `plugins`-folder

```
#!/opt/homebrew/bin/bash
# plugins

source "$CONFIG_DIR/colors.sh"

COUNT=$(tlmgr update --list --all --self | grep -c 'local:')
#sleep 5

COLOR=$RED

case "$COUNT" in
  [6-9][0-9]) COLOR=$RED
  ;;
  [3-5][0-9]) COLOR=$ORANGE
  ;;
  [1-2][0-9]) COLOR=$YELLOW
  ;;
  [1-9]) COLOR=$WHITE
  ;;
  0) COLOR=$GREEN
     COUNT=
  ;;
esac

sketchybar     \
  --set $NAME  \
  label=$COUNT \
  icon.color=$COLOR
```

and the file `tlmgr.sh` in the folder `items`:

```
#!/opt/homebrew/bin/bash
## items
#

tlmgr=(
  icon="" # <-- using Nerd-Font-Icon for TeX
  icon.font="$FONT:normal:30.0"
  label=?
  padding_right=15
    update_freq=20
  script="$PLUGIN_DIR/tlmgr.sh"
)

sketchybar --add event tlmgr_update \
           --add item tlmgr right   \
           --set tlmgr "${tlmgr[@]}" \
           --subscribe tlmgr tlmgr_update
```

Bonus points: I also like to get the list of outdated packages when I mouse over the TeX icon. Same goes for the outdated packages of homebrew. But I couldnt figure that out.

**Comment by bustinbung** [comment]
Not sure if you figured out the issue with the numbers showing up incorrectly (maybe check if sketchybar can run `tlmgr`; PATH issues?), but to get a list of packages on hover, you could use popup menus with the `mouse.*` events. I've added some pseudo-code below, you'd have to go through the [documentation](https://felixkratz.github.io/SketchyBar/config/popups) for more specifics. Hope this helps!

```sh
# item.sh #

tlmgr=(
  # item configuration

  # see documentation for more info
  popup.[topmost | blur_radius | background.* | ...]=value
)

sketchybar --add item tlmgr  \
           ...               \
           --subscribe tlmgr mouse.entered mouse.exited
```

```sh
# plugin.sh #

update() {
  # assuming this includes each package on its own line a la brew list -1
  PACKAGES=$(tlmgr get outdated packages)
  # counts number of lines in PACKAGES via here-string
  # (see: https://askubuntu.com/questions/678915/whats-the-difference-between-and-in-bash)
  COUNT=$(wc -l <<< $PACKAGES)

  for (( i = 0; i < COUNT; i++ )); do
    details=(
      # gets ith row in PACKAGES
      label="$(sed "${i}q;d" <<< $PACKAGES)" # from
               # https://stackoverflow.com/questions/6022384/bash-tool-to-get-nth-line-from-a-file
      # other item configuration
    )

    # see documentation for more info
    sketchybar --add item "tlmgr.outdatedpack.${i}" popup.tlmgr
  done
}

# I've found that I typically get better reliability when using one function
# for both mouse events and checking the event within the function instead
# of having one function for each event (especially when flicking over it).
# Experiment with it and see which works better for you.
mouse() {
  if [[ "$SENDER" == "mouse.entered" ]]; then
    sketchybar --set tlmgr popup.drawing=on
  else
    sketchybar --set tlmgr popup.drawing=off
  fi

}

# Example for one function per event
# mouse_enter() {
#   sketchybar --set tlmgr popup.drawing=on
# }

case "$SENDER" in
  "mouse.entered" | "mouse.exited") mouse ;;
  *) update ;;
esac
```

**Reply by ghost** [reply]
Thank you very much for the proposed solution. How do I find out if sketchybar can run `tlmgr`? In my "normal bash" I can run `tlmgr`. 

Unfortunately I cannot get your code working (yet). Need to dive deeper into the mouse-event issue. I added your code to the plugin-file however nothing happens when I click on the `TeX`-icon.

**Reply by bustinbung** [reply]
> How do I find out if sketchybar can run `tlmgr`?

The easiest way is probably to just run `sketchybar` from the command line and see what it outputs. If `sketchybar` can't run `tlmgr` you'll get an error along the lines of "`command not found`"

> I added your code to the plugin-file however nothing happens when I click on the TeX-icon.

If the script and mouse events are functioning properly, you could use `sketchybar --query` to see if there are any properties that are preventing the popup from showing. Documentation for `--query` is [here](https://felixkratz.github.io/SketchyBar/config/querying). Maybe check for opacity, width/height, content, etc.?
